xla/service/collective_pipeliner_test.cc

```diff
@@ -2228,5 +2229,90 @@ ENTRY main.3813_spmd {
+  get-tuple-element.397 = bf16[1,8,2048,32768]{3,2,1,0} get-tuple-element(param), index=2
+  get-tuple-element.398 = bf16[1,64,2048,32768]{3,2,1,0} get-tuple-element(param), index=3
+  get-tuple-element.399 = bf16[1,64,2048,32768]{3,2,1,0} get-tuple-element(param), index=4
+
+  constant.1 = bf16[] constant(2)
+  broadcast.3593 = bf16[1,8,2048,32768]{3,2,1,0} broadcast(constant.1), dimensions={}
+
+  add.2 = bf16[1,8,2048,32768]{3,2,1,0} add(broadcast.3593, get-tuple-element.395)
+  add.3 = bf16[1,8,2048,32768]{3,2,1,0} add(broadcast.3593, get-tuple-element.397)
+
+  all-gather.1 = bf16[1,64,2048,32768]{3,2,1,0} all-gather(add.2), channel_id=1, dimensions={1}, replica_groups={}
+  all-gather.2 = bf16[1,64,2048,32768]{3,2,1,0} all-gather(add.3), channel_id=2, dimensions={1}, replica_groups={}
+
+  mul.1 = bf16[1,64,2048,32768]{3,2,1,0} multiply(all-gather.1, get-tuple-element.398)
+  mul.2 = bf16[1,64,2048,32768]{3,2,1,0} multiply(all-gather.2, get-tuple-element.399)
+  add.4 = bf16[1,64,2048,32768]{3,2,1,0} add(mul.1, mul.2)
+
+  slice.1 = bf16[1,8,2048,32768]{3,2,1,0} slice(add.4), slice={[0:1], [0:8], [0:2048], [0:32768]}
+  slice.2 = bf16[1,8,2048,32768]{3,2,1,0} slice(add.4), slice={[0:1], [8:16], [0:2048], [0:32768]}
+  constant.2 = s32[] constant(1)
+  add.230 = s32[] add(get-tuple-element.394, constant.2)
+
+  ROOT tuple = (s32[], bf16[1,8,2048,32768]{3,2,1,0}, bf16[1,8,2048,32768]{3,2,1,0}, bf16[1,64,2048,32768]{3,2,1,0}, bf16[1,64,2048,32768]{3,2,1,0}) tuple(add.230, slice.1, slice.2, mul.1, mul.2)
+}
+
+ENTRY entry {
+  c0 = s32[] constant(0)
+  p0 = bf16[1,8,2048,32768]{3,2,1,0} parameter(0)
+  p1 = bf16[1,8,2048,32768]{3,2,1,0} parameter(1)
+  p2 = bf16[1,64,2048,32768]{3,2,1,0} parameter(2)
+  p3 = bf16[1,64,2048,32768]{3,2,1,0} parameter(3)
+
+  tuple = (s32[], bf16[1,8,2048,32768]{3,2,1,0}, bf16[1,8,2048,32768]{3,2,1,0}, bf16[1,64,2048,32768]{3,2,1,0}, bf16[1,64,2048,32768]{3,2,1,0}) tuple(c0, p0, p1, p2, p3)
+  while = (s32[], bf16[1,8,2048,32768]{3,2,1,0}, bf16[1,8,2048,32768]{3,2,1,0}, bf16[1,64,2048,32768]{3,2,1,0}, bf16[1,64,2048,32768]{3,2,1,0}) while(tuple), condition=while_cond, body=while_body
+  ROOT gte1 = bf16[1,8,2048,32768]{3,2,1,0} get-tuple-element(while), index=1
+}
+)";
+
+  auto module = ParseAndReturnUnverifiedModule(hlo_string, config_).value();
+
+  EXPECT_TRUE(
+      RunOptimizer(
+          module.get(), /*last_run=*/true, 0,
+          /*pipeline_use_tree=*/false,
+          /*process_different_sized_ops=*/false,
+          /*direction=*/CollectivePipeliner::PipeliningDirection::kBackward,
+          /*should_process=*/IsAllGather,
+          /*acceptable_formatting=*/HloPredicateTrue,
+          /*reuse_pipelined_op_buffer=*/HloPredicateTrue,
+          /*should_allow_loop_variant_parameter_in_chain=*/HloPredicateTrue,
+          /*postprocess_backward_peeled=*/std::nullopt,
+          /*postprocess_backward_rotated=*/std::nullopt,
+          /*should_add_loop_invariant_op_in_chain=*/true)
```

Comment:
Can we also assert that this does not happen if `should_add_loop_invariant_op_in_chain` is false?

Reply:
sure, let me change it.